File: server/internal/curation/domain/combination.go
package domain

// CombinationResponse is a historical recommendation, never a checkout quote.
// Versions fence Cart application; listing prices are not a catalog cache.
type CombinationResponse struct {
	SchemaVersion    string            `json:"schemaVersion"`
	ID               string            `json:"id"`
	Kind             string            `json:"kind"`
	Items            []CombinationItem `json:"items"`
	Currency         string            `json:"currency"`
	MinimumMinor     *int64            `json:"minimumMinor"`
	MaximumMinor     *int64            `json:"maximumMinor"`
	BudgetMinor      *int64            `json:"budgetMinor"`
	DifferenceMinor  *int64            `json:"differenceMinor"`
	BudgetStatus     string            `json:"budgetStatus"`
	Estimated        bool              `json:"estimated"`
	PriceBasis       string            `json:"priceBasis"`
	MissingTargetIDs []string          `json:"missingTargetIds"`
	SourceState      string            `json:"sourceState"`
	BudgetVersion    int64             `json:"budgetVersion"`
	CartVersion      int64             `json:"cartVersion"`
	CriteriaVersions map[string]int64  `json:"criteriaVersions"`
	Compatibility    string            `json:"compatibility"`
	Reasons          []string          `json:"reasons"`
	Tips             []CombinationTip  `json:"tips"`
	Cautions         []string          `json:"cautions"`
	BudgetAdvice     string            `json:"budgetAdvice"`
}

type CombinationItem struct {
	Ref                  string `json:"ref"`
	TargetID             string `json:"targetId"`
	CandidateID          string `json:"candidateId"`
	VariantID            string `json:"variantId,omitempty"`
	ConfigurationVersion int64  `json:"configurationVersion"`
	Quantity             int    `json:"quantity"`
	CartItemID           string `json:"cartItemId,omitempty"`
	CheckoutEligible     bool   `json:"checkoutEligible"`
	AllocationFit        string `json:"allocationFit"`
}

type CombinationTip struct {
	Label string `json:"label"`
	Body  string `json:"body"`
}

File: web/src/products/curation/domain/combination.ts
import type { ThreadResponse } from "./thread";
export type CombinationItem = { ref: string; targetId: string; candidateId: string; variantId?: string; configurationVersion: number; quantity: number; cartItemId?: string; checkoutEligible: boolean; allocationFit: "WITHIN" | "OVER" | "UNKNOWN" };
export type Combination = { schemaVersion: "vitlane.combination.v1"; id: string; kind: string; items: CombinationItem[]; currency: string; minimumMinor: number | null; maximumMinor: number | null; budgetMinor: number | null; differenceMinor: number | null; budgetStatus: "WITHIN" | "OVER" | "UNKNOWN" | "NO_LIMIT"; estimated: boolean; priceBasis: "EXACT" | "RANGE" | "ESTIMATED" | "UNKNOWN"; missingTargetIds: string[]; sourceState: string; budgetVersion: number; cartVersion: number; criteriaVersions: Record<string, number>; compatibility: "SUPPORTED" | "UNVERIFIED" | "UNRELATED" | "CONFLICT"; reasons: string[]; tips: { label: string; body: string }[]; cautions: string[]; budgetAdvice: string };
export type CombinationStatus = { schemaVersion: "vitlane.combination-status.v1"; current: boolean; reasonCode?: string; cartVersion: number };
export type CombinationCartMode = "ADD" | "REPLACE";

/** Present structured advice as ordinary reply paragraphs, preserving the existing conversation UI. */
export function combinationReplyText(response: ThreadResponse): string {
 const plan = response.combination;
 if (!plan) return response.body;
 const paragraphs = [
  response.body,
  [...plan.reasons, ...plan.tips.map(tip => tip.body)].filter(Boolean).join(" "),
  plan.cautions.filter(Boolean).join(" "),
  plan.budgetAdvice,
 ];
 return [...new Set(paragraphs.map(text => text.trim()).filter(Boolean))].join("\n\n");
}
